Add database

ArsenyIshtubaev opened the pull request.

ArsenyIshtubaev committed:
Спринт 10
-спроектировать ER диаграмму

ArsenyIshtubaev committed:
Спринт 10
-спроектировать ER диаграмму

ArsenyIshtubaev committed:
Спринт 10 (2)
-спроектировать ER диаграмму

ArsenyIshtubaev committed:
Спринт 10 (2)
-спроектировать ER диаграмму

ArsenyIshtubaev committed:
Спринт 10 (2)
-спроектировать ER диаграмму

ArsenyIshtubaev committed:
Спринт 10 (2)
-спроектировать ER диаграмму

ArsenyIshtubaev committed:
Спринт 11
-создать таблицы
-внести данные в таблицу

ArsenyIshtubaev committed:
Спринт 11
-добавить функциональность — сохранение состояния данных после перезапуска

ArsenyIshtubaev committed:
Спринт 11
-добавить функциональность — сохранение состояния данных после перезапуска

ArsenyIshtubaev committed:
Спринт 11
-добавить функциональность — сохранение состояния данных после перезапуска

DyndaNE commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,150 @@
+public class FilmDbStorage implements FilmStorage{
+
+    private final JdbcTemplate jdbcTemplate;
+    private final GenreDbStorage genreDbStorage;
```
одно хранилище не должно зависеть от другого хранилища
обращаться к жанрам следует через сервис  вызывать в нем методы хранилища жанров

DyndaNE commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,150 @@
+                null,
+                null
+                );
+            String sql = "select * from MPA where MPA_ID = ?";
```
зачем делать к мпа отдельный запрос если ты мог выполнить сразу джойн при получении фильма

DyndaNE commented on src/main/java/ru/yandex/practicum/filmorate/storage/FriendDbStorage.java:
```diff
@@ -0,0 +1,75 @@
+@Component
+public class FriendDbStorage implements FriendStorage{
+    private final JdbcTemplate jdbcTemplate;
+    private final UserDbStorage userDbStorage;
```
опять  же одно хранилище не может зависеть от другого хранилища

DyndaNE commented on src/main/java/ru/yandex/practicum/filmorate/storage/FriendDbStorage.java:
```diff
@@ -0,0 +1,75 @@
+        String sqlQuery1 = "select * from USERS where USER_ID = ?";
+        List<User> users1 = jdbcTemplate.query(sqlQuery1, userDbStorage::makeUser, userId);
+        if (users1.size() != 1) {
+            throw new StorageException("Пользователя с таким userId нет в базе данных");
```
проверки и исключения должны быть только в сервисе, вся бизнес логика должна быть в нём

DyndaNE commented on src/main/java/ru/yandex/practicum/filmorate/storage/FriendDbStorage.java:
```diff
@@ -0,0 +1,75 @@
+                "join (select * from FRIENDSHIP where FRIENDSHIP.USER_ID = ? " +
+                "AND FRIENDSHIP.FRIEND_STATUS = true) as fr2 "+
+                "ON fr1.FRIEND_ID = fr2.FRIEND_ID";
+        List<User> users = jdbcTemplate.query(sqlQuery, userDbStorage::makeUser, userId, friendId); */
```
ты прав лучше подумать над поиском друзей одним запросом

DyndaNE commented on src/main/java/ru/yandex/practicum/filmorate/storage/LikeDbStorage.java:
```diff
@@ -0,0 +1,57 @@
+        List<User> users = jdbcTemplate.query(sqlQuery2, userDbStorage::makeUser, userId);
+        if (users.size() != 1) {
+            throw new StorageException("Пользователя с таким id нет в базе данных");
+        }
```
опять же проверки и исключения должны быть в сервисе

ArsenyIshtubaev committed:
Спринт 11 (2)
-убрать зависимость хранилищ друг от друга, сделать вызов методов через сервисы

ArsenyIshtubaev committed:
Спринт 11 (2)
-убрать зависимость хранилищ друг от друга, сделать вызов методов через сервисы

DyndaNE commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,59 @@
+                                     "MPA_NAME" varchar(5) NOT NULL
+);
+
+CREATE TABLE IF NOT EXISTS "FILMS" (
```
кстати обычно в наименовании таблиц используют единственное число, а не множественное
но это мелочи

DyndaNE commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -93,21 +65,35 @@ public void validate(User user) throws ValidationException {
+                "AND FRIENDSHIP.FRIEND_STATUS = true) as fr2 "+
+                "ON fr1.FRIEND_ID = fr2.FRIEND_ID";
+        List<User> users = jdbcTemplate.query(sqlQuery, userDbStorage::makeUser, userId, friendId);
+        return users1;
```
а откуда тут запрос из хранилища? )
